### PR TITLE
add go module example

### PR DESCRIPTION
**Pull request by zeroshade**

### CODE REVIEW

**Comment by zeroshade** [comment]
@tifflhl Here you go! 

Oh yea, I should add the relevant github workflow for it too!

**Review by tifflhl — APPROVED** [review]

**Comment by tifflhl** [comment]
@zeroshade One more thing, I think all the .go files should have license headers?

**Comment by zeroshade** [comment]
@tifflhl which file is missing one? i thought i got them all?

**Comment by zeroshade** [comment]
@tifflhl any updates?

**Review by tifflhl — APPROVED** [review]

File: go/example.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	}
+	log.Println("[INFO] GetSchema was successful.")
+
+	schema, err := flight.DeserializeSchema(sc.GetSchema(), memory.DefaultAllocator)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Println("[INFO] Schema:", schema)
+
+	// Get the FlightInfo message to retireve the ticket corresponding to the query result set
```
Comment (kylepbit) [inline go/example.go:132]: Small Nit, there's a typo: retireve instead of retrieve

Comment (zeroshade) [reply go/example.go]: there ya go :)

**Comment by zeroshade** [comment]
@kylepbit  @tifflhl  any updates on getting this merged?

**Comment by stoffeastrom** [comment]
@zeroshade 👋  Hi again :D.

Do you have any pointers in adding support for connection with the go client to [dremio cloud](https://docs.dremio.com/cloud/arrow-flight/#sample-arrow-flight-client-applications) like in the [python](https://github.com/dremio-hub/arrow-flight-client-examples/blob/main/python/example.py) example?

**Comment by zeroshade** [comment]
It looks like dremio cloud uses a set-cookie response header to utilize a specific cookie for auth from a bearer token. You should be able to pretty easily use the custom middleware functionality in the v7 go arrow library to implement this by checking the headers (you can look at the flight integration tests for examples if you want).

I don't have a Dremio Cloud instance set up yet to test with, though you could probably follow the logic in the cookie middleware in the python example you linked.

If I have some time I can give it a try myself in the next week. Or if you put a PR together I can give it a review :)

**Comment by jduo** [comment]
@stoffeastrom @zeroshade ,
Dremio's session management will be based on HTTP cookies instead of implicit based on the bearer token starting in 21.0.0 and Dremio cloud. The old mode that's based on the bearer token will still be available though, and the upgrade process will use the old mode by default.

You'll also need to be able to supply an Authorization: Bearer header for Dremio cloud to supply the OAuth token.

**Comment by stoffeastrom** [comment]
@jduo I got some time today to see if the `python` example worked. Followed the `readme` but I can't get it to connect.

```sh
❯ python3 example.py -query 'select 1' -host 'data.eu.dremio.cloud' -port 443 -pat $DREMIO_PAT
[INFO] Authentication skipped until first request
[INFO] Query:  select 1
[ERROR] Exception: OSError('gRPC returned unavailable error, with message: failed to connect to all addresses. Detail: Unavailable')
Traceback (most recent call last):
  File "<redacted>/arrow-flight-client-examples/python/example.py", line 260, in <module>
    connect_to_dremio_flight_server_endpoint(args.hostname, args.port, args.username, args.password,
  File "<redacted>/arrow-flight-client-examples/python/example.py", line 236, in connect_to_dremio_flight_server_endpoint
    schema = client.get_schema(flight_desc, options)
  File "pyarrow/_flight.pyx", line 1300, in pyarrow._flight.FlightClient.get_schema
  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status
OSError: gRPC returned unavailable error, with message: failed to connect to all addresses. Detail: Unavailable
```

Something obvious I'm missing here?

**Comment by jduo** [comment]
@stoffeastrom , you'll also need to enable TLS to connect to Dremio cloud. Please add the -tls flag.

**Comment by stoffeastrom** [comment]
This got me a bit further. However it's now hanging and doesn't give back any results. 🤔 

```sh
❯ python3 example.py -tls -query 'select 1' -host 'data.eu.dremio.cloud' -port 443 -pat $DREMIO_PAT
[INFO] Enabling TLS connection
[INFO] Trusted certificates provided
[INFO] Authentication skipped until first request
[INFO] Query:  select 1
[INFO] GetSchema was successful
[INFO] Schema:  <pyarrow._flight.SchemaResult object at 0x11cc1d9d0>
[INFO] GetFlightInfo was successful
[INFO] Ticket:  <Ticket b'\n\x08select 1\x12"\n \n\x08select 1\x10\x04\x1a\x12\t\xd8\xbb\x19\xc7\x93\xa7\xd9\x1d\x11\x00\x88\xa0\x01\x7f\x13?\x95'>
```

**Comment by mcraealex-bq** [comment]
> This got me a bit further. However it's now hanging and doesn't give back any results. 🤔
> 
> ```shell
> ❯ python3 example.py -tls -query 'select 1' -host 'data.eu.dremio.cloud' -port 443 -pat $DREMIO_PAT
> [INFO] Enabling TLS connection
> [INFO] Trusted certificates provided
> [INFO] Authentication skipped until first request
> [INFO] Query:  select 1
> [INFO] GetSchema was successful
> [INFO] Schema:  <pyarrow._flight.SchemaResult object at 0x11cc1d9d0>
> [INFO] GetFlightInfo was successful
> [INFO] Ticket:  <Ticket b'\n\x08select 1\x12"\n \n\x08select 1\x10\x04\x1a\x12\t\xd8\xbb\x19\xc7\x93\xa7\xd9\x1d\x11\x00\x88\xa0\x01\x7f\x13?\x95'>
> ```

Thank you for reporting this. Can I ask how long you are waiting? Also, are you seeing a timeout message?

**Comment by omida-bitquill** [comment]
@stoffeastrom If the engines have to start for this query to run there may be a delay. If you go to the Dremio web UI and click on the "Jobs" (3 bullet points and lines) you will see what the status of the query is. It may take up to a minute or two for an engine to start.

**Comment by stoffeastrom** [comment]
@mcraealex-bq @omida-bitquill thank you for your swift replies. I didn't think of checking the jobs. Looking there now and I can see that they actually end up there. Looks like it's `unavailable` for some reason. I can try to query some samples instead.

image

**Comment by omida-bitquill** [comment]
> @mcraealex-bq @omida-bitquill thank you for your swift replies. I didn't think of checking the jobs. Looking there now and I can see that they actually end up there. Looks like it's `unavailable` for some reason. I can try to query some samples instead.
> 
> image

I believe the "unavailable" is referring to the Dataset, which would not exist in this case since you were doing a "SELECT 1". If you do it on one of the sample datasets/tables, it should have a respective dataset name.

**Pull request merged**